datalad_catalog/workflows.py

```diff
@@ -0,0 +1,248 @@
+    - The parent dataset already contains the subdataset commit
+    """
+    # Install super and subdatasets
+    super_ds = Dataset(superds_path)
+    sub_ds = Dataset(subds_path)
+    # Dataset workflow for super+subdataset
+    yield dataset_workflow(super_ds, catalog)
+    yield dataset_workflow(sub_ds, catalog)
+    # Set super dataset of catalog
+    catalog.main_id = super_ds.id
+    # - sync possible adjusted branch
+    super_ds.repo.localsync()
+    # - account for possibility of being on adjusted branch:
+    catalog.main_version = super_ds.repo.get_hexsha(
+        super_ds.repo.get_corresponding_branch()
+    )
+    catalog.set_main_dataset()
+
+
+def dataset_workflow(ds: Dataset, catalog, **kwargs):
+    """Run a dataset-specific catalog generation workflow.
+
+    This includes:
+      - dataset- and file-level metadata extraction
+      - extracted metadata translation
+      - adding translated metadata to a catalog"""
+    # 1. Run dataset-level extraction
+    extracted_file = Path(ds.path) / "extracted_meta.json"
+    for name in extractor_names_dataset:
+        if check_required_files(ds, name):
+            metadata_record = extract_dataset_level(ds, name)
+            write_jsonline_to_file(extracted_file, metadata_record)
+    # 2. Run file-level extraction, add output to same file
+    # - Not implemented yet
+    # 3. Run translation
+    translated_file = Path(ds.path) / "translated_meta.json"
+    with open(extracted_file) as file:
+        for line in file:
+            meta_dict = json.loads(line.rstrip())
+            try:
+                extr_name = meta_dict["extractor_name"]
+                extr_type = meta_dict["type"]
+                mapping_path = schema_dir / get_translation_map(
+                    extr_name, extr_type
+                )
+                write_jsonline_to_file(
+                    translated_file,
+                    translate_to_catalog(meta_dict, mapping_path),
+                )
+            except Exception as e:
+                raise (e)
+    # 4. Add translated metadata to catalog
+    return _add_to_catalog(catalog, translated_file, dict())
+
+
+def extract_dataset_level(dataset, extractor_name):
+    """Extract dataset-level metadata using metalad and a specific extractor"""
+    res = meta_extract(
+        extractorname=extractor_name,
+        dataset=dataset,
+        result_renderer="disabled",
+    )
+    return res[0]["metadata_record"]
+
+
+def conduct_extract_file_level(dataset):
+    """NOT IMPLEMENTED YET"""
+
+
+def extract_file_level(file):
+    """NOT IMPLEMENTED YET"""
+
+
+def check_required_files(dataset: Dataset, extractor_name: str):
+    for fn in required_files[extractor_name]:
+        f_path = Path(dataset.path) / fn
+        if not f_path.exists():
+            return False
+    return True
+
+
+def write_jsonline_to_file(filename, line):
+    """Write a single JSON line to file"""
+    open_as = "a"
+    with open(filename, open_as) as f:
```

Comment:
I think it could be written simply as:
```suggestion
    with open(filename, "a") as f:
```